rfc/curio-api-rfc.md

```diff
@@ -0,0 +1,157 @@
+3. IPNI announcements -- we plan to use IPNI announcements in a specific way with our pieces. Our understanding is the curio IPNI flow is in flux. We can try to integrate your IPNI api or just do it ourselves.
+
+## Basic flow
+
+In the proposal below, the basic flow is as follows;
+1. Create a proof set for Storacha on the SP (happens just once)
+2. Upload pieces from storacha with the piece storage API
+   - at this point, the piece is immediately retrievable but not being proven
+3. When enough pieces are received (128MB or more) create an aggregate root and add it to the proof set
+   - at this point, all pieces submitted to the proof set are retrievable AND proven
+
+## Proof sets API
+
+The following API describes how to create, read, update and delete proofsets managed by Curion. Essentially curio ill provide APIs that mirrored the Create/Add/Remove/Delete functions that will exist on chain for PDP proof sets, and then manage the submission of proofs on a schedule (I hear it's good at scheduling)
+
+### POST /proof-sets
+
+Create a new proof set for a specific 
+Request Body:
+```json
+{
+  // need to drill down on these propoerties
+}
+```
+
+Response:
+Code: 201
+Location header: "/proof-set/{set-id}"
+
+*TODO: do we need an interim response given this is a chain transaction with a place to fetch the set-id later?*
+
+### GET /proof-sets/{set-id}
+
+Response:
+Code: 200
+Body:
+```json
+{
+  "roots": [
+    // Root ids in proof set in order
+  ],
+  "ownerAddress": "f3...",
+  "challengePeriod": 15
+}
+```
+
+### POST /proof-sets/{set-id}/roots
+
+Append a root to the proof set, which may be an aggregation of one or more piece cids.
+
+Request Body:
+```json
+{
+  "rootCid": "bafy....root",
+  "pieces": [
+    {
+      "cid": "bafy...piece1",
+      "proof": [
+        "bafy...intermediate1",
+        "bafy...intermediate2"
+      ]
+    },
+    {
+      "cid": "bafy...piece2",
+      "proof": [
+        "bafy...intermediate3",
+        "bafy...intermediate4"
+      ]
+    },
+    //...
+  ],
+  "size": 1048576
+}
+```
+
+This API should fail if the all pieces were not previously stored with the Piece Storage API
+
+Response:
+Code: 201
+Location Header: "/proof-sets/{set id}/roots/{root id}
+
+### GET /proof-sets/{set id}/roots/{root id}
+
+Response Body:
+```json
+{
+  "rootCid": "bafy....root",
+  "pieces": [
+    {
+      "cid": "bafy...piece1",
+      "proof": [
+        "bafy...intermediate1",
+        "bafy...intermediate2"
+      ]
+    },
+    {
+      "cid": "bafy...piece2",
+      "proof": [
+        "bafy...intermediate3",
+        "bafy...intermediate4"
+      ]
+    },
+    //...
+  ],
+  "size": 1048576
+}
+```
+
+### DEL /proof-sets/{set id}/roots/{root id}
+
+Remove the given root id from the given proof set
+
+### DEL /proof-sets/{set id}
+
+Remove the specified proof set entirely
+
+
+## Piece Storage
+
+
+### POST /piece
```

Comment:
We definitely need to define how authorization works on this endpoint. This can't just be entirely open.

Also should define the lifecycle of the uploaded data somehow:
* How long is it expected to stick around in storage after upload before being included in a proof-set? When should the data be removed if not added to a proof set?
  * Signalling for expected indexing with IPNI / ipfs-type (trustless gateway/bitswap) retrievals, and who can retrieve the piece?
* What is the contract for retrieval - is it retrievable atomically when the notify hook is called? After inclusion in a proof set?

Reply:
Also what are the size bounds for pieces that you expect curio to support? We can support even very large pieces (100G+), but I don't think a client-push model is a good idea above 1GB, where managing short-term buffers becomes a real concern, and download retry becomes non-optional